Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+This document contains notes and is meant for planning the rollout of indexing changes, in particular removing spectrum number and detector IDs from `ISpectrum`.
+
+## Remarks
+
+It has been discussed more than once that the term *workspace index* is misleading.
+The only argument for keeping it is consistency.
+Throughout this document the better term *spectrum index* or simply *index* will be used.
+The nomenclature in the Mantid code base could be changed together with this major change.
+Note that this does not necessarily imply that the nomenclature on user-interface side changes as well.
+
+## Statistics
+
+Most relevant methods that need to be removed are part of `ISpectrum`.
+The following table gives and estimate of the number of occurrences of the respective methods.
+
+| Method        | Occurrences | Files  |
+| ------------- |:-------------:|:-----:|
+| `ISpectrum::getSpectrumNo()` | 271 | 99 |
+| `ISpectrum::setSpectrumNo()` | 153 | 104 |
+| `ISpectrum::copyInfoFrom()` | 26 | 17 |
+| `ISpectrum::ISpectrum()` | ? | ? |
+| `ISpectrum::operator=()` | ? | ? |
+| `???DetectorID?()` | ~900 | ~300 |
+| `EventList::operator+=(const EventList &)` | ? | ? |
+
+An initial survey shows that almost half of the occurrences are for workspace initialization.
+Those cases are usually comparatively easy and straightforward to replace by the new style of workspace factory functions defined in `DataObjects/WorkspaceCreation.h` that use `IndexInfo` to create a workspace.
+
+Other significant fractions of the occurrences are made up by logging and exceptions and by tests that initialize individual spectra.
+Furthermore there are comparisons between the respective spectrum and detector information in two workspaces.
+
+
+## General goals
+
+- Quite a few algorithms take an input workspace, extract information about spectrum numbers and detector IDs, and initialize a new workspace based on those data. Example include `SumSpectra`, `DiffractionFocussing`, and `ExtractSpectra`.
+  One goal of the indexing library is to remove explicit uses of spectrum numbers and detector IDs in as much code as possible.
+  Since we need to refactor algorithms similar to those mentioned above, it is probably a good idea to do it properly right away.
+  That is, we should *not* just refactor the manipulation of spectrum numbers and detector IDs via `IndexInfo`.
+  Instead, we should provide functionality for grouping and extraction of spectra to create a new `IndexInfo`, without the need for explicit handling of spectrum numbers and detector IDs.
+
+  As an example, consider
+
+  ```cpp
+  namespace Indexing {
+  IndexInfo extract(const IndexInfo &, const SpectrumIndexSet &);
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
```

Comment:
@OwenArnold After having tried when implementing the fake `SpectrumInfo`: Changing this will be a really major effort. For example:

- Currently detector IDs are default generated.
- In many tests no instrument is set, or only a dummy instrument that does not have the corresponding detector IDs.
- Many algorithms (corresponding to those tests) attempt to get parameters (such as `isMasked`) for spectra, but skip if there is no detector.

I we now enforce correct detector IDs, the algorithm behavior will change, and thus the tests results. I have seen cases where this affects hard coded values, so it will be really tedious to fix this.

What I am saying is: Of course we still need to do this, but I am starting to think that it might be better to try and stage it more. That is, when first adding `IndexInfo`, we still allow bad IDs. Only in a later step we fix this.
I still need to figure out whether or not this approach will lead to other problems though.